pytest_fixtures/satellite_auth.py

```diff
@@ -0,0 +1,153 @@
+import copy
+
+from nailgun import entities
+from pytest import fixture
+
+from robottelo.config import settings
+from robottelo.constants import AUDIENCE_MAPPER
+from robottelo.constants import GROUP_MEMBERSHIP_MAPPER
+from robottelo.constants import LDAP_ATTR
+from robottelo.constants import LDAP_SERVER_TYPE
+from robottelo.datafactory import gen_string
+from robottelo.rhsso_utils import create_mapper
+from robottelo.rhsso_utils import get_rhsso_client_id
+from robottelo.rhsso_utils import run_command
+
+
+@fixture(scope='function')
+def auth_source(module_org, module_loc):
+    ldap_data = {
+        'ldap_user_name': settings.ldap.username,
+        'ldap_user_passwd': settings.ldap.password,
+        'base_dn': settings.ldap.basedn,
+        'group_base_dn': settings.ldap.grpbasedn,
+        'ldap_hostname': settings.ldap.hostname,
+    }
+    return entities.AuthSourceLDAP(
+        onthefly_register=True,
+        account=ldap_data['ldap_user_name'],
+        account_password=ldap_data['ldap_user_passwd'],
+        base_dn=ldap_data['base_dn'],
+        groups_base=ldap_data['group_base_dn'],
+        attr_firstname=LDAP_ATTR['firstname'],
+        attr_lastname=LDAP_ATTR['surname'],
+        attr_login=LDAP_ATTR['login_ad'],
+        server_type=LDAP_SERVER_TYPE['API']['ad'],
+        attr_mail=LDAP_ATTR['mail'],
+        name=gen_string('alpha'),
+        host=ldap_data['ldap_hostname'],
+        tls=False,
+        port='389',
+        organization=[module_org],
+        location=[module_loc],
+    ).create()
```

Comment:
Function scoped fixtures - but no cleanup of the auth source after configuration?

Please consider two things here:
1. Module scoping the fixture for optimization - how many of the auth tests need function scope? What's the cost of setup/teardown of the auth sources
2. Teardown of the auth source - how will these fixtures behave when we're running the entire auth test suite? Are we relying on a behavior in `nailgun.entities.create()` to resolve the fact that we might try to configure the same auth source more than once?

Reply:
> Function scoped fixtures - but no cleanup of the auth source after configuration?
> 
> Please consider two things here:
> 
> 1. Module scoping the fixture for optimization - how many of the auth tests need function scope? What's the cost of setup/teardown of the auth sources
All of them need the function scope because we update the test with minor changes with same user from auth source. While if have already login with that user next test would use the same the auth source. This refactor has earlier similar issues because of that stability of tests wasn't there. Hence we added a check to remove everything based on test wise. Creating auth source is from API won't take that much time. 
If I wanted to mimic the same customer behaviour we have only a single auth source in the customer. If we make them both LDAP AD and IPA as module-level then end up with issue connecting two servers.         
      
> 2. Teardown of the auth source - how will these fixtures behave when we're running the entire auth test suite? Are we relying on behavior in `nailgun.entities.create()` to resolve the fact that we might try to configure the same auth source more than once?

This was the wrapper I added to make sure that everything gets clean up. If you think I should be placed below code some common location I can do that.    
https://github.com/omkarkhatavkar/robottelo/blob/cc306ff69f93737abb15c6a7b7dfc2498452efdb/tests/foreman/ui/test_ldap_authentication.py#L149-L168

Reply:
For #2 - I see. Yes, this should be in a function scoped fixture that looks like the following.  I would still consider making the function more specific to the ldap sources provided by the fixtures used for the test.  That way the test function stack is doing setup, test execution, and teardown all via linked fixtures.

Considering your responses on the session scoped ldap cleanup fixture below, sounds like you agree with making a change here in the direction I'm suggesting.

The session scoped ldap cleanup fixture below would only do user cleanup _before_ the tests run, not after or between any specific tests.

```
def ldap_cleanup():
    yield
    <iterate over auth sources and delete users>
```

Reply:
did you mean to remove all that decorator logic and get back to the fixture which will do all that cleanup? that fixture would be at function scope as the requirement is for the function scope? @mshriver is this my understanding is correct of your comment ?

Reply:
Correct, the behavior is actually built into pytest, I think we should use those existing mechanisms, not write our own decorators that re-implement behavior that pytest supports.

Reply:
added here 
https://github.com/SatelliteQE/robottelo/pull/7858/commits/79f816ca7e15e8d2cec4733b767b0655621d4b56#diff-ecddf5d413467e9d1edbc2cd0c22c984R116